proposals/bip-002.md

```diff
@@ -0,0 +1,99 @@
+
+This BIP proposes a mechanism to mark nodes with two possible completion status
+roles. This will make the UAST user have a wider view of the completeness of the
+information provided by it on specific nodes and also could enable new uses cases
+(whose specific definition is outside of the scope of this BIP) like helping debug
+drivers and provide a completeness summary, with the added benefit of providing a
+numerically observable measure of the completeness of every developed driver.
+
+## Example
+
+If a version of the UAST specification didn't have a role that could be matched
+for a "coroutine" type function supported by some language, the UAST could still
+mark those nodes as "FunctionDef", but without additional information, this would
+be providing incomplete and even misleading information to UAST users that could
+treat coroutines differently (for example when making stack call deep analysis).
+
+With this BIP implemented, the driver's programmer can add the role `Incomplete`
+to these coroutine nodes and the UAST user for that language could then look at
+the native AST properties to complete their semantic understanding of the node.
+
+## Specification
+
+### UAST Roles
+
+This change will add two new roles the UAST definition:
+
+- `Incomplete`: For nodes that have been annotated annotations doesn't express the
+  full semantic meaning.
+
+- `Unannotated`: For nodes that have received no other annotations.
+
+After this proposal have been fully implemented it'll be implicitly understood
+that nodes that have roles set and no `Incomplete` role are expressing full
+semantic information.
+
+### Driver and SDK annotation
+
+The Incomplete roles will be set explicitly by the driver's programmer using the
+normal process in he driver's `annotation.go` file.
+
+The Unannotated role will be added automatically by the SDK to nodes that didn't
+get any other role.
+
+
+## Impact
+
+- The new roles would have to be added to the SDK's `uast.go` file and the drivers
+  rebuild against it.
+
+- The `Incomplete` mark would have to be added by the driver developers when
+  defining the rules in the `annotation.go` file. Existing drivers will need to be
+  updated.
+
+- The SDK annotation process will implement the automatic would have to implement
+  setting the `Unannotated` marks. This will be implemented in the
+  `protocol/native/objecttonoder.go` file.
+
+- We'll add a new process to the build system that will check both the
+  `driver/normalizer/??ast/??ast.go` file with the static list of nodes for
+  existing but unannotated nodes and the integration test results and output a
+  summary of incomplete an unannotated nodes.
+
+- The documentation for the driver developer would have to be updated to include 
+  the purpose and mechanics of these new property/s or role/s. This is in the file 
+  `driver/annotations.md`.
+
+- This BIP will be added to the official documentation in the `proposals`
```

Comment:
Maybe not added by *published* (like in showing in http://doc.bblf.sh), will change it.